{
  State outcome=ERROR;
  if (ctx.hasTransaction()) {
    if (err.status().code().classification().rollbackTransaction() || ctx.implicitTransaction) {
      try {
        ctx.currentTransaction.failure();
        ctx.currentTransaction.close();
      }
 catch (      Throwable t) {
        ctx.log.error("While handling '" + err.status() + "', a second failure occurred when "+ "rolling back transaction: "+ t.getMessage(),t);
      }
 finally {
        ctx.currentTransaction=null;
      }
    }
 else {
      outcome=RECOVERABLE_ERROR;
    }
  }
  ctx.error(err);
  return outcome;
}
